modules/nlu/src/backend/application/index.ts

```diff
@@ -0,0 +1,124 @@
+import { NLU } from 'botpress/sdk'
+import _ from 'lodash'
+
+import { IBotFactory } from './bot-factory'
+import { IBotService } from './bot-service'
+import { BotNotMountedError } from './errors'
+import { ITrainingQueue } from './training-queue'
+import { ITrainingRepository } from './training-repo'
+import { Predictor, BotConfig, TrainingSession, TrainingState, TrainingId } from './typings'
+
+export class NLUApplication {
+  private _queueTrainingOnBotMount: boolean
+
+  constructor(
+    private _trainingQueue: ITrainingQueue,
+    private _engine: NLU.Engine,
+    private _botFactory: IBotFactory,
+    private _botService: IBotService,
+    queueTrainingOnBotMount: boolean = true
+  ) {
+    this._queueTrainingOnBotMount = queueTrainingOnBotMount
+  }
+
+  public async initialize() {
+    await this._trainingQueue.initialize()
+  }
+
+  public get trainRepository(): ITrainingRepository {
+    return this._trainingQueue.service.repository
+  }
+
+  public teardown = async () => {
+    await this._trainingQueue.teardown()
+    for (const botId of this._botService.getIds()) {
+      await this.unmountBot(botId)
+    }
+  }
+
+  public getHealth() {
+    return this._engine.getHealth()
+  }
+
+  public async getTraining(botId: string, language: string): Promise<TrainingState> {
+    return this._trainingQueue.getTraining({ botId, language })
+  }
+
+  async getAllTrainings(): Promise<TrainingSession[]> {
+    return this._trainingQueue.getAllTrainings()
+  }
+
+  async pauseTrainings() {
+    return this._trainingQueue.pause()
+  }
+
+  async resumeTrainings(): Promise<void> {
+    await this._trainingQueue.resume()
+  }
+
+  public hasBot = (botId: string) => {
+    return !!this._botService.getBot(botId)
+  }
+
+  public getBot(botId: string): Predictor {
+    const bot = this._botService.getBot(botId)
+    if (!bot) {
+      throw new BotNotMountedError(botId)
+    }
+    return bot
+  }
+
+  public mountBot = async (botConfig: BotConfig) => {
+    const { id: botId, languages } = botConfig
+    const { bot, defService, modelRepo } = await this._botFactory.makeBot(botConfig)
+    this._botService.setBot(botId, bot)
+
+    const makeDirtyModelHandler = (cb: (trainId: TrainingId) => Promise<void>) => async (language: string) => {
+      const latestModelId = await defService.getLatestModelId(language)
```

Comment:
probably something I don't understand but by reading it doesn't make sens that it's the job of a `definitionService` to return the latest model ID.  Do we have a `modelService` ? isn't it the logical thing to do with all the other service/repo components ?

Reply:
well, I had the same feeling.

We have a modelService but the modelId is only influenced by thoses 4 variables:

1. definitions // the definition service has this
2. languageCode // the bot config has this
3. engine specs // the engine has this
4. random seed // the bot config has this

So that's pretty much why its the definitions service job. But I have no strong feeling about this (it could be any module's resposnsibilty)